### PR TITLE
CLOUDP-67056: Implement search API in the go client

### PR DESCRIPTION
**Pull request by gssbzn**
## Description

Add search support 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: mongodbatlas/search.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+	if groupID == "" {
+		return nil, nil, NewArgError("GroupID", "must be set")
+	}
+	if clusterName == "" {
+		return nil, nil, NewArgError("ClusterName", "must be set")
+	}
+	if dbName == "" {
+		return nil, nil, NewArgError("DatabaseName", "must be set")
+	}
+	if collName == "" {
```
Comment (robcarlan-mlab) [inline mongodbatlas/search.go:43]: [np] `collName` -> `collectionName` and `dbName` -> `databaseName` to make it more explicit?

File: mongodbatlas/search.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	var root []*SearchIndex
+	resp, err := s.Client.Do(ctx, req, &root)
+
+	return root, resp, err
+}
+
+// Get gets one Atlas Search index by its indexId.
+//
+// See more: https://docs.atlas.mongodb.com/reference/api/fts-indexes-get-one/
+func (s *SearchServiceOp) GetIndex(ctx context.Context, projectID, clusterName, indexID string) (*SearchIndex, *Response, error) {
```
Comment (robcarlan-mlab) [inline mongodbatlas/search.go:69]: [np] Api docs reference group id rather than project id for these endpoints (the above one is group id though)

**Comment by gssbzn** [comment]
@themantissa ready for you

File: mongodbatlas/search.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+// SearchService provides access to the search related functions in the Atlas API.
+//
+// See more: https://docs.atlas.mongodb.com/reference/api/atlas-search/
+type SearchService interface {
+	ListIndexes(context.Context, string, string, string, string, *ListOptions) ([]*SearchIndex, *Response, error)
+	GetIndex(context.Context, string, string, string) (*SearchIndex, *Response, error)
+	CreateIndex(context.Context, string, string, *SearchIndex) (*SearchIndex, *Response, error)
+	UpdateIndex(context.Context, string, string, string, *SearchIndex) (*SearchIndex, *Response, error)
+	DeleteIndex(context.Context, string, string, string) (*Response, error)
+	ListAnalyzers(context.Context, string, string, *ListOptions) ([]*SearchAnalyzer, *Response, error)
```
Comment (themantissa) [inline mongodbatlas/search.go:22]: Curious if leaving Update Analyzers out was intentional? (https://docs.atlas.mongodb.com/reference/api/fts-analyzers-update-all/)

Comment (gssbzn) [reply mongodbatlas/search.go:22]: Somewhat intentional, there are plans to deprecate Custom Analyzers so I'm trying to limit us to only read them (if ever needed)

File: mongodbatlas/search.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+	Database       string        `json:"database"`
+	IndexID        string        `json:"indexID,omitempty"`
+	Mappings       *IndexMapping `json:"mappings,omitempty"`
+	Name           string        `json:"name"`
+	SearchAnalyzer string        `json:"searchAnalyzer,omitempty"`
+}
+
+// IndexMapping containing index specifications for the collection fields.
+type IndexMapping struct {
+	Dynamic bool                   `json:"dynamic"`
```
Comment (PacoDw) [inline mongodbatlas/search.go:216]: it's not better if it's a `*bool`?

**Pull request merged**